Rich Text Fields

stedmanblake opened the pull request:
- [x] Add RTF support to all the models:
  - [x] suggestion,
  - [x] experience (both on the _form and card/_form partials),
  - [x] idea (description, impact, implementation),
  - [x] recipe (description, evidence, and protips)
- [x] Make sure all `.description` are suffixed with `.html_safe`

---
### Suggestion
#### `http://localhost:3000/users/1`
## ![image](https://cloud.githubusercontent.com/assets/4050144/15003823/d1e22564-1164-11e6-820b-81bd36e14020.png)
#### `http://localhost:3000/` form
## ![image](https://cloud.githubusercontent.com/assets/4050144/15003894/b8e00c4c-1165-11e6-9a6e-e0829eda800b.png)
#### `http://localhost:3000/` show
## ![image](https://cloud.githubusercontent.com/assets/4050144/15003906/eaf1a1f0-1165-11e6-8751-ae38b58e352f.png)
### Experience
#### `experiences/show`
## ![image](https://cloud.githubusercontent.com/assets/4050144/15006039/ea3d75d4-1180-11e6-927f-a1892ca2d2e0.png)
#### `experiences/_form`
## ![image](https://cloud.githubusercontent.com/assets/4050144/15006068/835af502-1181-11e6-95a3-40fed723e681.png)
#### `users/cards/_submission`
## ![image](https://cloud.githubusercontent.com/assets/4050144/15006338/b223802c-1184-11e6-8ba6-fdfe77a55318.png)
#### `experiences/cards/_show`

![image](https://cloud.githubusercontent.com/assets/4050144/15006448/d11d0f2e-1185-11e6-823f-dc63b0cced01.png)

---
### Idea
#### `ideas/_form`
## ![image](https://cloud.githubusercontent.com/assets/4050144/15007152/78a850d2-118b-11e6-853e-1a8f312dc3ff.png)
#### `ideas/card/_show`
## ![image](https://cloud.githubusercontent.com/assets/4050144/15007136/40faad24-118b-11e6-98d9-c8512dfa4735.png)
#### `ideas/show`

![image](https://cloud.githubusercontent.com/assets/4050144/15007145/6844826a-118b-11e6-8dff-ad18d59b48f0.png)

---
### Recipe
#### `recipes/_form`

![image](https://cloud.githubusercontent.com/assets/4050144/15007380/629ff45a-118d-11e6-9860-c9e860851314.png)

---
#### `recipes/show`

![image](https://cloud.githubusercontent.com/assets/4050144/15007527/a95e483c-118e-11e6-8b3b-a82116f18d34.png)

---
#### `recipes/show`

![image](https://cloud.githubusercontent.com/assets/4050144/15007618/472315f2-118f-11e6-92b5-92c1a8367db1.png)

---
#### `users/cards/_submission`

![image](https://cloud.githubusercontent.com/assets/4050144/15007570/f0f689e8-118e-11e6-8dfd-931852d26d25.png)

stedmanblake committed:
Change Suggestion description fields to class: redactor. Update all suggestion.description areas to render html.

stedmanblake committed:
Change Experience form description fields to class: redactor.

stedmanblake committed:
Update experience.title to strip html from the description. Render html formatted submission title.

stedmanblake committed:
Update experience.title to strip html from the description. Render html formatted submission title.

stedmanblake committed:
Render HTML in experiences/card/show.

stedmanblake committed:
Change Idea form's description, impact, and implementation fields to RTF

stedmanblake committed:
Change Recipe form's description, evidence, and protips fields to RTF

stedmanblake commented:
@krnjn this is also ready for your review now!

krnjn commented on app/models/experience.rb:
```diff
@@ -48,7 +48,8 @@ class Experience < ActiveRecord::Base
   def title
-    self.description.present? ? self.description.truncate(50) : nil
+    # Since we allow RTF in the description, we should strip any HTML from the title
+    self.description.present? ? ActionController::Base.helpers.strip_tags(self.description).truncate(30) : nil
```
Any reason you went down to 30 from 50 chars? Otherwise this seems reasonable to me.

stedmanblake replied:
It just looked a little long relative to the objects with _actual_ titles that are rendered in the same list (on user#show). Happy to bump back up if you prefer!

krnjn commented on app/views/experiences/card/_show.html.erb:
```diff
@@ -10,7 +10,7 @@
       <% if experience.embed.present? %>
         <div class='col-sm-6'>
-          <p><%= experience.description %></p>
+          <p><%= experience.description.html_safe %></p>
```
Probably need to do this sort of thing on `app/views/experiences/show.html.erb` as well?

stedmanblake replied:
Yeah, that was already taken care of! Must've been my GitHub guardian OctoAngel 👼

krnjn commented on app/views/suggestions/_preview.html.erb:
```diff
@@ -7,9 +7,9 @@
       <div class='media-body'>
         <h4 class='media-heading text-default'><%= suggestion.title %></h4>
-        <p class='media-description'><%= truncate(suggestion.description, length: 250) %></p>
+        <p class='media-description'><%= truncate(suggestion.description.html_safe, length: 250) %></p>
```
Probably want to strip tags on this as well: `truncate(strip_tags(suggestion.description).html_safe, length: 250)`

krnjn replied:
Yeah looks like we were already doing this in the other `_preview` files with `truncate_html` – here's an example: `<p class='media-description'><%= truncate_html(idea.description, length: 250).html_safe %></p>`

stedmanblake replied:
Got it - good call

krnjn commented on app/views/users/cards/_submission.html.erb:
```diff
@@ -5,10 +5,10 @@
-      <i class='fa <%= submission.icon %>'></i> 
-      <%= link_to submission.title, after_update_object_path_for(submission) %>
+      <i class='fa <%= submission.icon %>'></i>
+      <%= link_to submission.title.html_safe, after_update_object_path_for(submission) %>
```
Don't think we'd need `.html_safe` on title since we don't allow for RTF on title fields? Even with the experience one you stripped tags, so we should be fine here. Generally `.html_safe` is not actually that safe as if it's `nil`, it will cause the page to error out, so we should use it sparingly and only when we're certain that input will be present.

stedmanblake replied:
Ah yes, I'd forgotten I stripped tags in `experience.title` thanks!

krnjn commented on app/views/users/cards/_submission.html.erb:
```diff
@@ -5,10 +5,10 @@
+      <%= link_to submission.title.html_safe, after_update_object_path_for(submission) %>
     </h4>
-    <p class='zero-margin-bottom'><%= submission.description %></p>
+    <p class='zero-margin-bottom'><%= submission.description.html_safe if submission.description.present? %></p>
```
I think we always require description with every submission, so we can avoid the extra check here for `if submission.description.present?` – know that I suggested it, but think we'll be fine without it.

stedmanblake replied:
Sounds good!

krnjn commented:
The screenshots are looking good! Thanks for including those – loving the fake text inputs haha. 

Also, just looked and it looks like we've included the `truncate_html` gem and helper which may be of use to us instead of stripping tags and truncating in the views – https://github.com/hgmnz/truncate_html. Want to give that a shot with the few spots we're using `.truncate` or `truncate()` in the views?

stedmanblake committed:
Strip tags from suggestion desctiption in the preview view

stedmanblake committed:
Remove html_save from submission.title; Experience - the only possible contender for html in the title - strips any html in its title method

stedmanblake committed:
Remove html_save from submission.title; Experience - the only possible contender for html in the title - strips any html in its title method

stedmanblake committed:
Remove unnecessary check for description's existence

krnjn committed:
Add RichTextPresence Validator

**What**
This commit introduces some behavior to help validate our new rich text field inputs on the models/attributes that we introduce rich text. Added validator, specs, and calls to it in models. Also added some styling updates so that rich text box turns same "error" red as other input error states. Finally, namespaced validator specs to avoid clashing / persisted error messages.

krnjn commented:
@stedmanblake – this was looking good, I added a custom validator for the rich text presence (otherwise an empty redactor field would post `<p><br><p>` and it would be considered "valid" since it was present). Did you have any other work on it? I had to force push after rebasing changes on master, so it got a little bit messy. If not, I think this was looking good to go!

stedmanblake commented:
@krnjn sounds good - thanks for the update!
